### PR TITLE
[WIP] implements ParameterGroupsModuleMixin

### PR DESCRIPTION
**Pull request by lgvaz**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Starts to implement #2005  
The idea of this PR is to make the model `param_groups` more accessible, so they can be quickly used to freeze layers and to setup differential learning rates.

### Examples
For a resnet34 it would look something like this:
```python
def model_splits(self):
    model = resnet34(False) # Defined elsewhere, I just put it here as an example
    groups = []
    groups += nn.Sequential(model.conv1, model.bn1)
    groups += [layer for name,layer in model.named_children() if name.startswith("layer")]
    groups += [model.fc] # Considering we already switched the head
    return groups
```

So if we want to train only the last layers (i.e. the head in this case):
```python
model.freeze_to(-1)
```

If we want to make the entire network trainable with differential learning rates
```python
model.freeze_to(0) # Or model.unfreeze
n_groups = len(list(model.param_splits()))
lrs = np.linspace(1e-5, 1e-3, n_groups).tolist()
param_groups = model.get_optimizer_param_groups(lrs)
# Pass param groups to the optimizer
```

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by lgvaz** [comment]
I added an example so we don't have to keep depending on the colab notebook. This does not means that the example will be present on the final PR (but maybe in a future PR? 😉)

**Comment by lgvaz** [comment]
Currently the user has to manually define the learning rate for all PGs, as we see done here:
```python
n_param_groups = 2
freeze_lrs = [1e-3] * n_param_groups
unfreeze_lrs = np.linspace(5e-6, 5e-4, n_param_groups).tolist()
```

This is quite a lot of boilerplate code that can be avoided.
Fastai solves this by allowing the user to pass a `slice`, and then it interpolates the values for all PGs.

What do you all think? Suggestions?

**Comment by williamFalcon** [comment]
This is a nice start! I think i see what you're trying to do now.
I think that this particular approach:

1. is too specific
2. hides away pytorch details
3. is not general enough.

Sounds like what we're looking for is a new LR Scheduler...

```python
i'm not sure about adding these custom functions to the lightning module.

1. We always want to use pytorch and not wrap it or somehow obscure it.
2. this means the user has to remember more stuff about the lightning module.

this all seems to be wrapped up nicely by a learning_rate scheduler

```python
class FineTuneScheduler(Scheduler):
    def __init__(self, schedules):
        self.schedules = schedules

    def get_lr(self):
       step_num = self.last_epoch
       for schedule in self.schedules:
             if schedule['epoch'] == step_num:
                # do the thing
```

Then to start it it would be something like this:
```python
scheduler = FineTuneScheduler([
   {'params': [nn.Sequential(self.c_d1, self.c_d1_bn), self.c_d2], 'action': 'freeze', 'epoch' 0},
   {'params': [self.c_d2], 'action': 'unfreeze', 'epoch' 2},
])


```

**Comment by Borda** [comment]
@lgvaz Examples are always good lol

**Comment by lgvaz** [comment]
I updated the OP and added an example, hopefully the intentions of this PR are clearer.

The intentions of this PR are not about the "transfer learning phase switch flow", but only about parameter groups and differential learning rates. It got confusing because of the example file I included (which deals with the training loop).

Let's forget about the training loop on this one and figure out how to properly design it in #2005 and #2006, as William suggested

**Comment by lgvaz** [comment]
Something I would like some feedback on are these lines:
```python
n_groups = len(list(model.param_splits()))
lrs = np.linspace(1e-5, 1e-3, n_groups).tolist()
param_groups = model.get_optimizer_param_groups(lrs)
```

Currently if the user wants to define differential lrs, he/she has to manually specific a lr for each param group.

Should we implement an option where the user just pass the lower and the highest LR and we internally interpolate it to all layers? it would look something like this:
```python
param_groups = model.get_optimizer_param_groups(slice(1e-5, 1e-3))
```

**Comment by Borda** [comment]
> For a resnet34 it would look something like this:
> 
> ```python
> def model_splits(self):
>     model = resnet34(False) # Defined elsewhere, I just put it here as an example
>     groups = []
>     groups += nn.Sequential(model.conv1, model.bn1)
>     groups += [layer for name,layer in model.named_children() if name.startswith("layer")]
>     groups += [model.fc] # Considering we already switched the head
>     return groups
> ```

just thinking a case when you have a trained model which was already trained and you saved the write could it be simply loaded with this updated model with groups? just for some back-compatibility...
Also maybe an option could be to pace a binary vector with layer association and cuntom turn on/off (unfreeze/freeze) any of them...

**Comment by Borda** [comment]
I quite like Will's suggestion 1 and do not know now any better way...
anyway, it would be nice to have it in next release to get some user feedback on it :]

**Comment by williamFalcon** [comment]
yes. let’s do it as schedulers

**Comment by lgvaz** [comment]
Sure! I'll work on that this weekend =)

**Comment by lgvaz** [comment]
@Borda I was not quite able to fully understand your question

Basically you're looking for a way of automatic saving the parameter groups without having to define that cited piece of code?

**Comment by Borda** [comment]
> Basically you're looking for a way of automatic saving the parameter groups without having to define that cited piece of code?

I was more thinking about some back-compatibility...
let me draw a use-case: imagine you have a model defined without these groups and you trained this model for a few days, now you want to update it with the fine-tuning schema so you update the model definition to have these groups and now the question is can you simply load past trained weights?

**Comment by lgvaz** [comment]
Aaah I understand the question now

Yes, you would be able to load the weights previous to these changes

`model_splits` is only called once you need your parameter groups, we don't actually touch the original model

So you could even train a model with some defined parameter groups, save it, change `model_splits`, and load it again

**Comment by lgvaz** [comment]
So, I started to look back at this, there is a annoying issue implementing this the way William described here:
```python
# unfreeze head after 1 epoch
sched = FineTuneScheduler(nn.Sequential(self.c_d1, self.c_d1_bn), self.c_d2, 1)
```

Let's say we also want to change the learning rates at this point, so we could pass `lrs=[1,2]`  

This would internally create a new scheduler with this new learning rates for each parameter group. The problem is that the learning rates should match the parameter groups we passed when creating the optimizer. But when creating the optimizer we did not specified any parameter groups, so the only way that I currently see of doing this is to recreate the optimizer with the new parameter groups.

That involves being extra careful with the optimizer state, we don't want to lose any stats when switching phases, so that would probably mean we would need to save the state of the old optimizer and reload on this new one (but I'm not sure how this will play out with the new parameter groups).

This also generates the confusion of, what happens in the following scenario:

```python
sched = FineTuneScheduler(nn.Sequential(self.c_d1, self.c_d1_bn), self.c_d2, 1, lrs=[.1,.2])
sched = FineTuneScheduler(self.c_d1_bn, self.c_d2, 1, lr=[.2, .1])
```

What happened is that we broke a previous parameter group into a separate part. First question is, what should happen to the lr of `self.c_d1`? Probably stay the same (.1)?

From the engineering side, this is quite complicated, we have to identify that a previous parameter group was broken into smaller pieces so we don't mistakenly add both parts in the optimizer.

Or maybe we define that only the groups passed to the scheduler are trainable, that eliminates the previous issue of broken parameter groups.

---

At the end, allowing the user to pass arbitrary layers to the schedulers generates a lot of complexity, and I don't see any real added benefit over pre-defining the parameter groups before creating the optimizer (so they can already added to the optimizer at creation time), similar to what I suggested in the original post.

I think that will create much simpler code =)

**Comment by Kshitij09** [comment]
What's the progress of this PR? is there anything I could help with?

cc: @lgvaz

**Comment by lgvaz** [comment]
Hello @Kshitij09!

It has been a while since I worked on this, as you can see I had some problems implementing William suggestions, but haven't heard back since

Instead of continuing working on this PR, I think a better strategy would be to start clean from master and open a new PR implementing the `Scheduler` as William suggested

I'll actually close this one since it's really outdated.
